### PR TITLE
git: Instrument calls to ResolveRevision

### PR DESCRIPTION
**Pull request by ryanslade**
We see a huge number of `rev-parse` commands being sent to gitserver and
we have a hunch that this is the main culprit.

### CODE REVIEW

File: internal/vcs/git/revisions.go [hunk 1]
```diff
@@ -47,6 +49,13 @@ type ResolveRevisionOptions struct {
 	NoEnsureRevision bool // do not try to fetch from remote if revision doesn't exist locally
 }
 
+const labelEnsureRevision = "ensure_revision"
+
+var resolveRevisionCounter = promauto.NewCounterVec(prometheus.CounterOpts{
+	Name: "src_resolve_revision_total",
+	Help: "The number of times we call internal/vcs/git/ResolveRevision",
+}, []string{labelEnsureRevision})
```
Comment (indradhanush) [inline internal/vcs/git/revisions.go:57]: Looks like `labelEnsureRevision` is used just once. Maybe we can directly use the string in this case?
```suggestion
}, []string{"ensure_revision"})
```

File: internal/vcs/git/revisions.go [hunk 2]
```diff
@@ -47,6 +49,13 @@ type ResolveRevisionOptions struct {
 	NoEnsureRevision bool // do not try to fetch from remote if revision doesn't exist locally
 }
 
+const labelEnsureRevision = "ensure_revision"
```
Comment (indradhanush) [inline internal/vcs/git/revisions.go:52]:
```suggestion
```

**Review by indradhanush — APPROVED** [review]
LGTM. Approve to unblock.

It wasn't clear to me from the description what is ResolveRevision a culprit of. :thinking:

**Comment by ryanslade** [comment]
> LGTM. Approve to unblock.
> 
> It wasn't clear to me from the description what is ResolveRevision a culprit of. 🤔

It's a culprit of make the majority of our `rev-parse` commands

**Comment by keegancsmith** [comment]
@ryanslade FYI check out honeycomb, we heavily sample it though due to the high volume. FYI the root cause is zoekt-indexserver polling sourcegraph to find out if a repo has changed. We are addressing this very soon.

**Comment by ryanslade** [comment]
> @ryanslade FYI check out honeycomb, we heavily sample it though due to the high volume. FYI the root cause is zoekt-indexserver polling sourcegraph to find out if a repo has changed. We are addressing this very soon.

Ah, that's good to know.

I'm looking at this from a slightly different angle to see if there's a way to avoid all the HA work in gitserver by instead exposing a simpler API to the rest of the system that could be implemented by a KV store. The KV store itself could then be an off the shelf solution that already supports HA.

Any thoughts on this approach?

**Comment by keegancsmith** [comment]
> I'm looking at this from a slightly different angle to see if there's a way to avoid all the HA work in gitserver by instead exposing a simpler API to the rest of the system that could be implemented by a KV store. The KV store itself could then be an off the shelf solution that already supports HA.
> 
> Any thoughts on this approach?

Some commands won't work well on the KV store, eg `git log`. But the most common commands (eg `rev-parse` and `show`) would work well. I think if you have rev parsing and showing a specific file/dir working, you solve 99% of request availability. Then for requests which require falling back to git they have partial avail. This sounds like a good trade-off, but I wonder at the complexity of having get a file in a KV store is. I think just rev-parse is likely not enough.

**Comment by ryanslade** [comment]
I was looking at `rev-parse` as I had a similar thought around covering the majority of the reads, not all of them, and `rev-parse` is our biggest culprit.

By KV store I'm also considering S3 like bucket storage, so something like Cloud Storage for .com and Minio for customer instances. I believe we are already using Minio for code-intel uploads.

**Comment by keegancsmith** [comment]
> I was looking at `rev-parse` as I had a similar thought around 
> covering the majority of the reads, not all of them, and 
> `rev-parse` is our biggest culprit. 

Expect the large majority of these to disappear (atleast for 
cloud).

My intuition is that our background requests (the majority) 
tolerate unavailability. So any addition of HA should be justified 
by improving user requests (or the vast majority of them).

FYI We log things like actor/internal/etc in honeycomb 
gitserver-exec dataset.

**Comment by ryanslade** [comment]
Do you have more details about the changes coming soon to reduce `rev-parse`?

**Comment by keegancsmith** [comment]
Not really. Our next work item (ie likely next week) is to change how zoekt interacts with sourcegraph from polling to push. It may involve changes to gitserver such that we can know when a repo changes, rather than polling.

**Pull request merged**